### ISSUE TITLE
Plugin ck-editor - can't install som gems

### ISSUE THREAD

**Issue by yozart** [docker-library/redmine#336]
Hello,

I'm migrating my old redmine 2.6 to a newer version (5.x). Migration by itself worked great. I can find my issues and config. Everything seems ok.
I need to import plugins now. The first one is already causing me some troubles. 

Plugin: CK-Editor FREE by RedmineX
https://www.redmine.org/plugins/redmine-x-ck-editor-free

Version of Docker Redmine tested :
- redmine:alpine
- redmine:latest
- redmine:[5.0.9]

Error message in docker:alpine
```
Your Gemfile lists the gem puma (>= 0) more than once.
You should probably keep only one of them.
Remove any duplicate entries and specify the gem only once.
While it's not a problem now, it could cause errors if you change the version of one of them later.
Bundler can't satisfy your Gemfile's dependencies.
Install missing gems with 'bundle install'.
Your Gemfile lists the gem puma (>= 0) more than once.
You should probably keep only one of them.
Remove any duplicate entries and specify the gem only once.
While it's not a problem now, it could cause errors if you change the version of one of them later.
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Installing mimemagic 0.3.10 with native extensions
Bundler::DirectoryRemovalError: Could not delete previous installation of
'/usr/local/bundle/gems/mimemagic-0.3.10'.
The underlying error was ArgumentError: parent directory is world writable,
Bundler::FileUtils#remove_entry_secure does not work; abort:
"/usr/local/bundle/gems/mimemagic-0.3.10" (parent directory mode 40777), with
backtrace:
/usr/local/lib/ruby/3.2.0/bundler/vendor/fileutils/lib/fileutils.rb:1387:in
'remove_entry_secure'
  /usr/local/lib/ruby/3.2.0/bundler.rb:332:in 'rm_rf'
/usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:111:in
'strict_rm_rf'
  /usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:19:in 'install'
  /usr/local/lib/ruby/3.2.0/bundler/source/rubygems.rb:202:in 'install'
  /usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:54:in 'install'
/usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:16:in
'install_from_spec'
/usr/local/lib/ruby/3.2.0/bundler/installer/parallel_installer.rb:156:in
```

Message with 5.0.9-bookworm
```
Unfortunately, an unexpected error occurred, and Bundler cannot continue.
First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=parent+directory+is+world+writable%2C+Bundler++FileUtils%23remove_entry_secure+does+not+work%3B+abort++%22%2Fusr%2Flocal%2Fbundle%2Fgems%2Fmimemagic-0.3.10%22+%28parent+directory+mode+40777%29&type=Issues
If there aren't any reports for this error yet, please fill in the new issue form located at https://github.com/rubygems/rubygems/issues/new?labels=Bundler&template=bundler-related-issue.md, and copy and paste the report template above in there.
```

Does anyone managed to install it ? What shall I do ?

**Comment by whalelines** [comment]
We are looking into a longer-term fix, but for now you could run the following command prior to trying to install the plugin:

```sh
chmod -R o-w "$GEM_HOME"/gems/*
```

**Comment by yozart** [comment]
Hello Whalelines,
Thanks for the comment. Can you tell me how/where to add this line ?

I'm using portainers stacks for creating the container, here is my docker compose :

```
redmine:
    image: redmine:alpine
    container_name: redmine
    restart: always
    environment:
      REDMINE_DB_ADAPTER: "sqlserver"
      REDMINE_DB_DATABASE: "Test"
      REDMINE_DB_SQLSERVER: "xx.xx.xx.xx"
      REDMINE_DB_PORT: "1433"
      REDMINE_DB_USERNAME: "username"
      REDMINE_DB_PASSWORD: "password"
      VIRTUAL_HOST: "sub.domain.com"
      REDMINE_PLUGINS_MIGRATE: true
    ports:
      - "3000:3000"
    volumes:
      - REDMINE_DATA:/usr/src/redmine/files
      - REDMINE_PLUGINS:/usr/src/redmine/plugins
    networks:
      - common_network
```

**Comment by LaurentGoderre** [comment]
Can you try:

```
command: ["bash", "-c", "chmod -R o-w \"$$GEM_HOME\"/gems/* && rails server -b 0.0.0.0"]
```

**Comment by yozart** [comment]
I'm getting : 
```
Please configure your config/database.yml first

/usr/local/lib/ruby/3.2.0/bundler/resolver.rb:304:in `raise_not_found!': Could not find gem 'kaminari' in locally installed gems. (Bundler::GemNotFound)
```

this is my docker compose now : 
```
redmine:
    image: redmine:alpine
    container_name: redmine
    restart: always
    environment:
      REDMINE_DB_ADAPTER: "sqlserver"
      REDMINE_DB_DATABASE: "Test"
      REDMINE_DB_SQLSERVER: "xx.xx.xx.xx"
      REDMINE_DB_PORT: "1433"
      REDMINE_DB_USERNAME: "username"
      REDMINE_DB_PASSWORD: "password"
      VIRTUAL_HOST: "sub.domain.com"
      REDMINE_PLUGINS_MIGRATE: true
    ports:
      - "3000:3000"
    volumes:
      - REDMINE_DATA:/usr/src/redmine/files
      - REDMINE_PLUGINS:/usr/src/redmine/plugins
    networks:
      - common_network
    command: ["bash", "-c", "chmod -R o-w \"$$GEM_HOME\"/gems/* && rails server -b 0.0.0.0"]
```

isn't it related to the gem folder available inside the plugin folder ?
/usr/src/redmine/plugins/redmine_ckeditor/gems (which contains only mimemagic-0.4.3.gem)

**Comment by LaurentGoderre** [comment]
Maybe this instead:

```
entrypoint: ["sh", "-c", "chmod -R o-w \"$$GEM_HOME\"/gems/* && /docker-entrypoint.sh && rails server -b 0.0.0.0"]
```

**Comment by tianon** [comment]
I'm looking into this now, and I think there's two parts here -- one is that the offending `chmod` we're doing in our `Dockerfile` is no longer necessary (at least in my testing -- no `--user` value I could come up with was broken after removing it, no matter which database type I chose).  I think it _would_ be reasonable to replace with something like `find "$GEM_HOME" -maxdepth 1 -type d -exec chmod 1777 '{}' +`, which would meet the spirit of the line without going quite as wide as what it's currently doing (essentially following the change in https://github.com/docker-library/ruby/pull/406 but in a less wide way than our current code), but I haven't checked whether the actual code of `Bundler::FileUtils#remove_entry_secure` would still balk at 1777 like it is currently balking at 777.

The other is that different plugins are going to have very different requirements, and it's pretty unreasonable for us to believe we can set up the image out-of-the-box for _all_ of them to work successfully without additional work.  For example, the one I tried (https://github.com/haru/redmine_logs) immediately tried to create a directory in `public/plugin_assets` -- after making that directory have 1777 permissions, it did work, but I don't know if that's reasonable for us to do by default.

In other words, I think that if there are reasonable/minor changes we can make to the images to make running with plugins in general easier, we should do them, but otherwise we need to pragmatically take the same stance as we do in the WordPress image: https://hub.docker.com/_/wordpress#:~:text=There%20are%20an%20infinite%20number%20of%20possible%20plugins%2C%20and%20they%20potentially%20require%20any%20extension%20PHP%20supports.%20Including%20every%20PHP%20extension%20that%20exists%20would%20dramatically%20increase%20the%20image%20size. (namely, that we cannot reasonably accommodate all use cases, and many are going to require custom setup code).

**Comment by yozart** [comment]
> entrypoint: ["sh", "-c", "chmod -R o-w \"$$GEM_HOME\"/gems/* && /docker-entrypoint.sh rails server -b 0.0.0.0"]

@LaurentGoderre thanks for your reply, I tested your proposal and got following errors : 

in redmine:latest & redmine:5.0.9
```
Installing mimemagic 0.3.10 with native extensions
Bundler::DirectoryRemovalError: Could not delete previous installation of
`/usr/local/bundle/gems/mimemagic-0.3.10`.
The underlying error was ArgumentError: parent directory is world writable,
Bundler::FileUtils#remove_entry_secure does not work; abort:
"/usr/local/bundle/gems/mimemagic-0.3.10" (parent directory mode 40777), with
backtrace:
/usr/local/lib/ruby/3.2.0/bundler/vendor/fileutils/lib/fileutils.rb:1387:in
`remove_entry_secure'
  /usr/local/lib/ruby/3.2.0/bundler.rb:332:in `rm_rf'
/usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:111:in
`strict_rm_rf'
  /usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:19:in `install'
  /usr/local/lib/ruby/3.2.0/bundler/source/rubygems.rb:202:in `install'
  /usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:54:in `install'
/usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/usr/local/lib/ruby/3.2.0/bundler/installer/parallel_installer.rb:156:in

`do_install'
/usr/local/lib/ruby/3.2.0/bundler/installer/parallel_installer.rb:147:in
`block in worker_pool'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:62:in `apply_func'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:57:in `block in process_queue'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:54:in `loop'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:54:in `process_queue'
/usr/local/lib/ruby/3.2.0/bundler/worker.rb:90:in `block (2 levels) in
create_threads'
Bundler Error Backtrace:
/usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:115:in `rescue in
strict_rm_rf'
/usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:110:in
`strict_rm_rf'
  /usr/local/lib/ruby/3.2.0/bundler/rubygems_gem_installer.rb:19:in `install'
  /usr/local/lib/ruby/3.2.0/bundler/source/rubygems.rb:202:in `install'
  /usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:54:in `install'
/usr/local/lib/ruby/3.2.0/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/usr/local/lib/ruby/3.2.0/bundler/installer/parallel_installer.rb:156:in

`do_install'
/usr/local/lib/ruby/3.2.0/bundler/installer/parallel_installer.rb:147:in
`block in worker_pool'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:62:in `apply_func'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:57:in `block in process_queue'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:54:in `loop'
  /usr/local/lib/ruby/3.2.0/bundler/worker.rb:54:in `process_queue'
/usr/local/lib/ruby/3.2.0/bundler/worker.rb:90:in `block (2 levels) in
create_threads'
An error occurred while installing mimemagic (0.3.10), and Bundler cannot

.
In Gemfile:
  paperclip was resolved to 6.1.0, which depends on
    mimemagic
```

@tianon I do agree with the general concept. you can't be compliant with every plugin for redmine but you should ensure that regular ones (I.E. the ones respecting the general rules for redmine) are working out of the box. (I don't know if CK_editor is respecting expected rules of plugins, from what I see, it seems that yes - but my knowledges are very limited in ruby).

**Comment by whalelines** [comment]
My initial suggestion was erroneous because gem is complaining about the parent directory, so the gems directory also needs the other-write bit removed. Can you try

```
entrypoint: ["sh", "-c", "chmod -R o-w \"$$GEM_HOME/gems\" && /docker-entrypoint.sh rails server -b 0.0.0.0"]
```

i.e., remove the wildcard character (and ensure the backslashes are maintained).

**Comment by yozart** [comment]
got a little bit farther, i had to remove other-write for folder extensions also

```
entrypoint: ["sh", "-c", "chmod -R o-w \"$$GEM_HOME/gems\" && chmod -R o-w \"$$GEM_HOME/extensions\" && /docker-entrypoint.sh rails server -b 0.0.0.0"]
```

Now I have the following error:
```
Your Gemfile lists the gem puma (>= 0) more than once.
You should probably keep only one of them.
Remove any duplicate entries and specify the gem only once.
While it's not a problem now, it could cause errors if you change the version of one of them later.
Bundler can't satisfy your Gemfile's dependencies.
Install missing gems with `bundle install`.
Your Gemfile lists the gem puma (>= 0) more than once.
You should probably keep only one of them.
Remove any duplicate entries and specify the gem only once.
While it's not a problem now, it could cause errors if you change the version of one of them later.
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Installing mimemagic 0.3.10 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
    current directory: /usr/local/bundle/gems/mimemagic-0.3.10/ext/mimemagic
/usr/local/bin/ruby -rrubygems /usr/local/bundle/gems/rake-13.2.1/exe/rake
RUBYARCHDIR\=/usr/local/bundle/extensions/x86_64-linux/3.2.0/mimemagic-0.3.10
RUBYLIBDIR\=/usr/local/bundle/extensions/x86_64-linux/3.2.0/mimemagic-0.3.10
rake aborted!
Could not find MIME type database in the following locations:
["/usr/local/share/mime/packages/freedesktop.org.xml",
"/opt/homebrew/share/mime/packages/freedesktop.org.xml",
"/opt/local/share/mime/packages/freedesktop.org.xml",
"/usr/share/mime/packages/freedesktop.org.xml"]
Ensure you have either installed the shared-mime-info package for your
distribution, or
obtain a version of freedesktop.org.xml and set FREEDESKTOP_MIME_TYPES_PATH to
      
the location
of that file.
This gem might be installed as a dependency of some bigger package, such as
rails, activestorage,
axlsx or cucumber. While most of these packages use the functionality of this
gem, some gems have
included this gem by accident. Set USE_FREEDESKTOP_PLACEHOLDER=true if you are
certain that you
do not need this gem, and wish to skip the inclusion of freedesktop.org.xml.
The FREEDESKTOP_PLACEHOLDER option is meant as a transitional feature, and will
be deprecated in
the next release.
/usr/local/bundle/gems/mimemagic-0.3.10/ext/mimemagic/Rakefile:15:in
`locate_mime_database'
/usr/local/bundle/gems/mimemagic-0.3.10/ext/mimemagic/Rakefile:39:in `block in
<top (required)>'
/usr/local/bundle/gems/rake-13.2.1/exe/rake:27:in `<main>'
Tasks: TOP => default
(See full trace by running task with --trace)
```

**Comment by whalelines** [comment]
It seems this game of [Whac-A-Mole](https://en.wikipedia.org/wiki/Whac-A-Mole) has moved beyond issues with the `redmine` DOI. Perhaps cd-editor is not that popular a plugin.

I would recommend reading the error carefully and then trying to address that.

For the current issue, it seems the relevant line is

```
Ensure you have either installed the shared-mime-info package for your distribution
```

For `redmine:latest`, this would mean adding `apt-get update && apt-get install -y shared-mime-info` to your entrypoint command before invoking `/docker-entrypoint.sh`. See https://www.debian.org/doc/manuals/debian-faq/pkgtools.en.html#apt-get .

**Comment by yozart** [comment]
It worked ! Thanks a lot for helping me solving this ! below the Docker compose I used:

```
redmine:
    image: redmine:latest
    container_name: redmine
    restart: always
    environment:
      REDMINE_DB_ADAPTER: "sqlserver"
      REDMINE_DB_SQLSERVER: "xx.xx.xx.xx"
      REDMINE_DB_PORT: "1433"
      REDMINE_DB_USERNAME: "username"
      REDMINE_DB_PASSWORD: "password"
      VIRTUAL_HOST: "sub.domain.com"
      REDMINE_PLUGINS_MIGRATE: true
    ports:
      - "3000:3000"
    volumes:
      - REDMINE_DATA:/usr/src/redmine/files
      - REDMINE_PLUGINS:/usr/src/redmine/plugins
    networks:
      - common_network
    entrypoint: ["sh", "-c", "chmod -R o-w \"$$GEM_HOME/gems\" && chmod -R o-w \"$$GEM_HOME/extensions\" && apt-get update && apt-get install -y shared-mime-info && /docker-entrypoint.sh rails server -b 0.0.0.0"]
```
